### PR TITLE
Align on Commonalities' subscription-model by using `sink` and `sinkCredentials`

### LINKED ISSUE #332: Alignment of notification web hook to CloudEvent subscription model
**Problem description**

Commonalities 0.4.0 introduced the CloudEvent subscription model with 'sink' and 'sinkCredentials' instead of 'notificationUrl' and 'notificationAuthToken', including the needed schema, within chapter 12.1 of the API Design Guidelines.
quality-on-demand uses implicit subscription for QoS Session and must be aligned for the Fall24 release.

**Expected behavior**

Align quality-on-demand API with the changes introduced by https://github.com/camaraproject/Commonalities/pull/198 ff.
Change org.camaraproject.qod.v0.qos-status-changed -> org.camaraproject.quality-on-demand.v0.qos-status-changed 

**Alternative solution**

**Additional context**

#299 has to be adapted accordingly

### PR DESCRIPTION
**Pull request by maxl2287**
#### What type of PR is this?

Add one of the following kinds:

* enhancement/feature

#### What this PR does / why we need it:
Aligns on Commonalities' updated subscription model.

#### Which issue(s) this PR fixes:
Fixes #332

### CODE REVIEW

File: code/API_definitions/quality-on-demand.yaml [hunk 1]
```diff
@@ -609,6 +602,99 @@ components:
...
+    SinkCredential:
+      type: object
+      properties:
+        credentialType:
+          type: string
+          enum:
+            - PLAIN
+            - ACCESSTOKEN
+            - REFRESHTOKEN
+          description: "The type of the credential."
```
Comment (hdamker) [inline code/API_definitions/quality-on-demand.yaml:614]: Commonalities 0.4.0 says in the Design Guidelines explicit: "Type of the credential - MUST be set to ACCESSTOKEN for now". So we should mentioned it also here:

```suggestion
          description: "The type of the credential. With the current API version the type MUST be set to `ACCESSTOKEN`"
```

File: code/API_definitions/quality-on-demand.yaml [hunk 2]
```diff
@@ -609,6 +602,99 @@ components:
...
+          properties:
+            accessToken:
+              description: REQUIRED. An access token is a previously acquired token granting access to the target resource.
+              type: string
+            accessTokenExpiresUtc:
+              type: string
+              format: date-time
+              description: REQUIRED. An absolute UTC instant at which the token shall be considered expired.
+            accessTokenType:
+              description: REQUIRED. Type of the access token (See [OAuth 2.0](https://tools.ietf.org/html/rfc6749#section-7.1)).
```
Comment (hdamker) [inline code/API_definitions/quality-on-demand.yaml:656]:
```suggestion
              description: REQUIRED. Type of the access token (See [OAuth 2.0](https://tools.ietf.org/html/rfc6749#section-7.1)). For the current version of the API the type MUST be set to `Bearer`.
```

File: code/API_definitions/quality-on-demand.yaml [hunk 3]
```diff
@@ -37,8 +37,9 @@ info:
     * **Duration**:
     Duration (in seconds) for which the QoS session (between application client and application server) should be created. Limits for session duration can be set by the implementation for the QoS profile. The user may request a termination before its expiration.
 
-    * **Notification URL and token**:
-    Developers may provide a callback URL on which notifications about all status change events of the session (eg. session termination) can be received from the service provider. This is an optional parameter.
+    * **Sink URL and credentials**:
+    Developers may provide a callback URL (`sink` param) on which notifications about all status change events of the session (eg. session termination) can be received from the service provider. This is an optional parameter.
+    Additionally, sink credentials may be provided to authenticate and authorize the delivery of events to the sink.
```
Comment (hdamker) [inline code/API_definitions/quality-on-demand.yaml:42]:
```suggestion
    Additionally, sink credentials may be provided to authenticate and authorize the delivery of events to the sink. In order to be updated in future this object is polymorphic. See detail below. It is RECOMMENDED for subscription consumer to provide credential to protect notification endpoint.
```

**Review by hdamker — CHANGES_REQUESTED** [review]
sinkCredentials are currently within the Design Guidelines quite restricted. Added here to the descriptions.

**Review by jlurien — COMMENTED** [review]
You have to correct as well:       bearerFormat: "{$request.body#/webhook/notificationAuthToken}" -> bearerFormat: "{$request.body#/sinkCredential.credentialType}"

**Comment by jlurien** [comment]
> sinkCredentials are currently within the Design Guidelines quite restricted. Added here to the descriptions.

In QoD Provisioning I have adjusted the previous text to:

```
    * **Notification URL and token**:
    Developers may provide a callback URL (`sink`) on which notifications about all status change events (eg. provisioning termination) can be received from the service provider. This is an optional parameter. The notification will be sent as a CloudEvent compliant message. If `sink` is included, it is RECOMMENDED for the client to provide as well the `sinkCredential` property to protect the notification enpoint. In the current version,`sinkCredential.credentialType` MUST be set to `ACCESSTOKEN` if provided.
```

cc @maxl2287

**Review by jlurien — COMMENTED** [review]
I have added also 410 as response to       
     callbacks:
        notifications:
          "{$request.body#/sink}":

As in the artifact because it has an explict reference to notifications:

    Generic410:
      description: Gone
      headers:
        x-correlator:
          $ref: "#/components/headers/x-correlator"
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/ErrorInfo"
          examples:
            GENERIC_410_GONE:
              description: Use in notifications flow to allow API Consumer to indicate that its callback is no longer available
              value:
                status: 410
                code: GONE
                message: Access to the target resource is no longer available.

File: code/API_definitions/quality-on-demand.yaml [hunk 4]
```diff
@@ -37,8 +37,9 @@ info:
     * **Duration**:
     Duration (in seconds) for which the QoS session (between application client and application server) should be created. Limits for session duration can be set by the implementation for the QoS profile. The user may request a termination before its expiration.
 
-    * **Notification URL and token**:
-    Developers may provide a callback URL on which notifications about all status change events of the session (eg. session termination) can be received from the service provider. This is an optional parameter.
+     * **Notification URL and token**:
+    Developers may provide a callback URL (`sink`) on which notifications about all status change events (eg. provisioning termination) can be received from the service provider. This is an optional parameter. The notification will be sent as a CloudEvent compliant message.
```
Comment (hdamker) [inline code/API_definitions/quality-on-demand.yaml:41]:
```suggestion
    Developers may provide a callback URL (`sink`) on which notifications about all status change events (eg. provisioning termination) can be received from the service provider. This is an optional parameter. The notification will be sent as a CloudEvent compliant message.
```

**Comment by hdamker** [comment]
> @hdamker shell I also update the BadRequest with the error-examples?

@maxl2287 the current Generic400 "BadRequest" is a copy from line 131ff of [Commonalities/blob/main/artifacts/CAMARA_common.yaml](https://github.com/camaraproject/Commonalities/blob/main/artifacts/CAMARA_common.yaml) and fits for the other endpoints.

INVALID_CREDENTIAL and INVALID_TOKEN are specific for createSession only - so I suppose you have to introduce and additional `CreateSessionBadRequest400` similar as in the [event-subscription-template](https://github.com/camaraproject/Commonalities/blob/9233f834c9cfb44dbb843ca72f22085f8559d4da/artifacts/camara-cloudevents/event-subscription-template.yaml#L799)? 

@jlurien also relevant for #299

**Review by jlurien — APPROVED** [review]

File: code/API_definitions/quality-on-demand.yaml [hunk 5]
```diff
@@ -952,6 +952,35 @@ components:
                 code: "INVALID_TOKEN"
                 message: "Only bearer token is supported"
 
+    GenericExtendSessionDuration400 :
```
Comment (hdamker) [inline code/API_definitions/quality-on-demand.yaml]:
```suggestion
    GenericExtendSessionDuration400:
```

**Review by hdamker — APPROVED** [review]

**Pull request merged**